Index: zaptel/zaptel.c
===================================================================
--- zaptel/zaptel.c	(revision 134)
+++ zaptel/zaptel.c	(working copy)
@@ -1327,3 +1327,7 @@
 #ifdef CONFIG_ZAPATA_NET
 MODULE_DEPEND (ztp, sppp, 1, 1, 1);
 #endif
+
+#ifdef ECHO_CAN_OSLEC
+MODULE_DEPEND(ztdev, oslec, 1, 1, 1);
+#endif
\ No newline at end of file
Index: zaptel/zconfig.h
===================================================================
--- zaptel/zconfig.h	(revision 134)
+++ zaptel/zconfig.h	(working copy)
@@ -80,8 +80,9 @@
 /* #define ECHO_CAN_MARK2 */
 /* #define ECHO_CAN_MARK3 */
 /* #define ECHO_CAN_KB1 */
+/* #define ECHO_CAN_MG2 */
 /* This is the new latest and greatest */
-#define ECHO_CAN_MG2
+#define ECHO_CAN_OSLEC
 
 /*
  * Uncomment for aggressive residual echo suppression under
Index: zaptel/zaptel.h
===================================================================
--- zaptel/zaptel.h	(revision 134)
+++ zaptel/zaptel.h	(working copy)
@@ -141,6 +141,15 @@
 #elif defined(ECHO_CAN_MG2)
 #define ZAPTEL_ECHO_CANCELLER "MG2"
 #include "mg2ec.h"
+/* Start Open Source Line Echo Canceller (OSLEC) -----------------*/
+#elif defined(ECHO_CAN_OSLEC)
+#define ZAPTEL_ECHO_CANCELLER "OSLEC"
+#include "oslec.h"
+#define echo_can_create oslec_echo_can_create
+#define echo_can_free oslec_echo_can_free
+#define echo_can_update oslec_echo_can_update
+#define echo_can_traintap oslec_echo_can_traintap
+/* End Open Source Line Echo Canceller (OSLEC) -------------------*/
 #else
 #define ZAPTEL_ECHO_CANCELLER "MARK3"
 #include "mec3.h"
Index: zaptel/oslec.h
===================================================================
--- zaptel/oslec.h	(revision 0)
+++ zaptel/oslec.h	(revision 0)
@@ -0,0 +1,48 @@
+/*
+  oslec.h
+  David Rowe
+  7 Feb 2007
+
+  Interface for OSLEC module.
+*/
+
+/*
+  Copyright (C) 2007 David Rowe
+ 
+  All rights reserved.
+ 
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License version 2, as
+  published by the Free Software Foundation.
+ 
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+ 
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+*/
+
+#ifndef __OSLEC__
+
+#define EC_FREE(a) free(a, M_DEVBUF)
+#define EC_MALLOC(size) malloc(size, M_DEVBUF, M_WAITOK)
+
+#include "arith.h"
+
+struct echo_can_state {
+  void *ec;
+};
+
+struct echo_can_state *oslec_echo_can_create(int len, int adaption_mode);
+void oslec_echo_can_free(struct echo_can_state *ec);
+short oslec_echo_can_update(struct echo_can_state *ec, short iref, short isig);
+int oslec_echo_can_traintap(struct echo_can_state *ec, int pos, short val);
+static inline void echo_can_init(void) {}
+static inline void echo_can_shutdown(void) {}
+short oslec_hpf_tx(struct echo_can_state *ec, short txlin);
+
+#endif
+

Property changes on: zaptel/oslec.h
___________________________________________________________________
Name: svn:executable
   + *

