--- ../../development-bundle/ui/ui.sortable.js.orig	2009-11-24 11:52:10.000000000 +0100
+++ ../../development-bundle/ui/ui.sortable.js	2009-11-24 11:53:03.000000000 +0100
@@ -364,7 +364,7 @@
 		var str = []; o = o || {};
 
 		$(items).each(function() {
-			var res = ($(o.item || this).attr(o.attribute || 'id') || '').match(o.expression || (/(.+)[-=_](.+)/));
+			var res = ($(o.item || this).attr(o.attribute || 'id') || '').match(o.expression || (/(.+)[_](.+)/));
 			if(res) str.push((o.key || res[1]+'[]')+'='+(o.key && o.expression ? res[1] : res[2]));
 		});
 
