Index: trunk/i4b/src/sys/dev/usb/ehci_pci.c
===================================================================
--- trunk/i4b/src/sys/dev/usb/ehci_pci.c	(revision 737)
+++ trunk/i4b/src/sys/dev/usb/ehci_pci.c	(working copy)
@@ -146,7 +146,7 @@
 	case 0x523910b9:
 		return "ALi M5239 USB 2.0 controller";
 
-	case 0x10227463:
+	case 0x74631022:
 		return "AMD 8111 USB 2.0 controller";
 
 	case 0x20951022:
