--- sys/boot/i386/libi386/bootinfo.c.orig	2006-02-14 07:20:15.000000000 +0100
+++ sys/boot/i386/libi386/bootinfo.c	2008-08-23 21:46:19.000000000 +0200
@@ -145,8 +145,8 @@
      * XXX: Note that until the kernel is ready to respect multiple consoles
      * for the boot messages, the first named console is the primary console
      */
-    if (!strcmp(string, "vidconsole"))
-	howto &= ~RB_SERIAL;
+    /*if (!strcmp(string, "vidconsole"))
+	howto &= ~RB_SERIAL;*/
 
     free(string);
 
