--- sys/kern/vfs_mount.c.orig	Tue Oct 24 17:02:39 2006
+++ sys/kern/vfs_mount.c	Sat Dec  9 16:35:45 2006
@@ -1541,7 +1541,7 @@
 		strcpy(path, ROOTNAME);
 
 	error = kernel_vmount(
-	    MNT_RDONLY | MNT_ROOTFS,
+	    MNT_ROOTFS,
 	    "fstype", vfsname,
 	    "fspath", "/",
 	    "from", path,
--- sys/i386/i386/vm_machdep.c.orig	Sat Dec  9 14:05:24 2006
+++ sys/i386/i386/vm_machdep.c	Sat Aug 18 15:10:27 2007
@@ -600,6 +600,12 @@
 		outl(0xcfc, 0xf);
 	}

+	if (strcmp(cpu_vendor, "AuthenticAMD") == 0 &&
+	    (cpu_id & 0xfff0) == 0x05a0) {
+		/* PC Engines ALIX reset via CS5536 DIVIL_SOFT_RESET */
+		wrmsr(0x51400017, 1LL);
+	}
+
 #ifdef PC98
 	/*
 	 * Attempt to do a CPU reset via CPU reset port.
