--- src.orig/libipsec/pfkey_dump.c	Wed Dec  7 14:45:05 2005
+++ src/libipsec/pfkey_dump.c	Sat Dec 23 23:08:00 2006
@@ -78,6 +78,9 @@
 #define SADB_X_EALG_RC5CBC	SADB_EALG_RC5CBC
 #endif
 #endif
+#if defined(SADB_X_EALG_AES) && ! defined(SADB_X_EALG_AESCBC)
+#define SADB_X_EALG_AESCBC  SADB_X_EALG_AES
+#endif
 
 #define GETMSGSTR(str, num) \
 do { \
--- src.orig/setkey/token.l	Wed Jun 29 15:01:30 2005
+++ src/setkey/token.l	Sat Dec 23 23:08:15 2006
@@ -84,6 +84,9 @@
 #ifndef SADB_X_EALG_AESCTR
 #define SADB_X_EALG_AESCTR	(-1)
 #endif
+#if defined(SADB_X_EALG_AES) && ! defined(SADB_X_EALG_AESCBC)
+#define SADB_X_EALG_AESCBC  SADB_X_EALG_AES
+#endif
 %}
 
 /* common section */
