--- Makefile.orig	Sat Nov  2 00:02:57 2002
+++ Makefile	Tue May 10 13:58:23 2005
@@ -14,10 +14,10 @@
 # http://www.openssl.org/  Make sure the SSL_TREE definition points to the
 # tree with your OpenSSL installation - depending on how you installed it,
 # it may be in /usr/local instead of /usr/local/ssl.
-#SSL_TREE =	/usr/local/ssl
-#SSL_DEFS =	-DUSE_SSL
-#SSL_INC =	-I${SSL_TREE}/include
-#SSL_LIBS =	-L${SSL_TREE}/lib -lssl -lcrypto
+SSL_TREE =	/usr
+SSL_DEFS =	-DUSE_SSL
+SSL_INC =	-I${SSL_TREE}/include
+SSL_LIBS =	-L${SSL_TREE}/lib -lssl -lcrypto
 
 
 BINDIR =	/usr/local/sbin
--- mini_httpd.c.orig	Wed Dec  3 19:27:22 2003
+++ mini_httpd.c	Sun Dec 18 11:39:28 2005
@@ -74,7 +74,7 @@
 
 
 #if defined(AF_INET6) && defined(IN6_IS_ADDR_V4MAPPED)
-#define USE_IPV6
+/* #define USE_IPV6 */
 #endif
 
 #ifndef STDIN_FILENO
@@ -141,7 +141,7 @@
 #define AUTH_FILE ".htpasswd"
 #endif /* AUTH_FILE */
 #ifndef READ_TIMEOUT
-#define READ_TIMEOUT 60
+#define READ_TIMEOUT 30
 #endif /* READ_TIMEOUT */
 #ifndef WRITE_TIMEOUT
 #define WRITE_TIMEOUT 300
@@ -167,13 +167,25 @@
 #endif /* USE_IPV6 */
     } usockaddr;
 
+typedef struct {
+	int			cpid;	/* child PID - 0 if unused */
+	in_addr_t	caddr;	/* client address */
+} conninfo;
 
 static char* argv0;
 static int debug;
 static unsigned short port;
+static conninfo* clients;
+static int maxproc;
+static int maxperip;
+static sigset_t sigchildset;
+static int currproc;
 static char* dir;
 static char* data_dir;
 static int do_chroot;
+static int captivemode;
+static char* cpelementpath;
+static char* cpelementhost;
 static int vhost;
 static char* user;
 static char* cgi_pattern;
@@ -209,6 +221,7 @@
 static size_t request_size, request_len, request_idx;
 static int method;
 static char* path;
+static char* captive_reqpath;
 static char* file;
 static char* pathinfo;
 struct stat sb;
@@ -322,9 +335,15 @@
     argv0 = argv[0];
     debug = 0;
     port = 0;
+	maxproc = 16 ;
+	maxperip = 0 ;
+	currproc = 0 ;
     dir = (char*) 0;
     data_dir = (char*) 0;
     do_chroot = 0;
+    captivemode = 0;
+    cpelementpath = NULL;
+    captive_reqpath = NULL;
     vhost = 0;
     cgi_pattern = (char*) 0;
     url_pattern = (char*) 0;
@@ -377,6 +396,20 @@
 	    ++argn;
 	    port = (unsigned short) atoi( argv[argn] );
 	    }
+	else if ( strcmp( argv[argn], "-maxproc" ) == 0 && argn + 1 < argc )
+	    {
+	    ++argn;
+	    maxproc = (unsigned short) atoi( argv[argn] );
+	    if (maxproc <= 0)
+	    	usage();
+	    }
+	else if ( strcmp( argv[argn], "-maxperip" ) == 0 && argn + 1 < argc )
+	    {
+	    ++argn;
+	    maxperip = (unsigned short) atoi( argv[argn] );
+	    if (maxperip < 0)
+	    	usage();
+	    }
 	else if ( strcmp( argv[argn], "-d" ) == 0 && argn + 1 < argc )
 	    {
 	    ++argn;
@@ -431,12 +464,24 @@
 	    ++argn;
 	    max_age = atoi( argv[argn] );
 	    }
+	else if ( strcmp( argv[argn], "-cpelement" ) == 0 && argn + 2 < argc )
+	    {
+	    ++argn;
+	    cpelementpath = argv[argn];
+	    ++argn;
+	    cpelementhost = argv[argn];
+	    }
+	else if ( strcmp( argv[argn], "-a" ) == 0 )
+	    captivemode = 1;
 	else
 	    usage();
 	++argn;
 	}
     if ( argn != argc )
 	usage();
+	
+	if (maxproc < maxperip)
+	    usage();
 
     cp = strrchr( argv0, '/' );
     if ( cp != (char*) 0 )
@@ -445,6 +490,16 @@
 	cp = argv0;
     openlog( cp, LOG_NDELAY|LOG_PID, LOG_DAEMON );
 
+	if (maxperip != 0) {
+		int i;
+		clients = e_malloc(sizeof(conninfo) * maxproc);
+		for (i = 0; i < maxproc; i++)
+			clients[i].cpid = 0;
+		
+		sigemptyset(&sigchildset);
+		sigaddset(&sigchildset, SIGCHLD);
+	}
+
     if ( port == 0 )
 	{
 #ifdef USE_SSL
@@ -722,6 +777,7 @@
 	    exit( 1 );
 	    }
 	/* Check for unnecessary security exposure. */
+	/*
 	if ( ! do_chroot )
 	    {
 	    syslog( LOG_WARNING,
@@ -729,6 +785,7 @@
 	    (void) fprintf( stderr,
 		"%s: started as root without requesting chroot(), warning only\n", argv0 );
 	    }
+	*/
 	}
 
     /* Catch various signals. */
@@ -751,6 +808,7 @@
 
     init_mime();
 
+	/*
     if ( hostname == (char*) 0 )
 	syslog(
 	    LOG_NOTICE, "%.80s starting on port %d", SERVER_SOFTWARE,
@@ -759,7 +817,8 @@
 	syslog(
 	    LOG_NOTICE, "%.80s starting on %.80s, port %d", SERVER_SOFTWARE,
 	    hostname, (int) port );
-
+	*/
+	
     /* Main loop. */
     for (;;)
 	{
@@ -816,7 +875,7 @@
 	    }
 	if ( conn_fd < 0 )
 	    {
-	    if ( errno == EINTR || errno == EAGAIN )
+	    if ( errno == EINTR || errno == EAGAIN || errno == ECONNABORTED )
 		continue;	/* try again */
 #ifdef EPROTO
 	    if ( errno == EPROTO )
@@ -827,6 +886,36 @@
 	    exit( 1 );
 	    }
 
+	/* If we've reached max child procs, then close the connection - 
+	   don't attempt to send back a response since that itself may 
+	   cause our process to hang. */
+	if (currproc >= maxproc) {
+		close(conn_fd) ;
+		continue ;
+	}
+	
+	sigprocmask(SIG_BLOCK, &sigchildset, NULL);
+	
+	/* If maxperip is enabled, count the number of existing connections
+	   from this client and close the connection if the max is exceeded. */
+	if (maxperip != 0) {
+		int i;
+		int nconns = 0;
+		
+		for (i = 0; i < maxproc; i++) {
+			if (clients[i].cpid == 0)
+				continue;
+			else if (clients[i].caddr == usa.sa_in.sin_addr.s_addr)
+				nconns++;
+		}
+		
+		if (nconns >= maxperip) {
+			close(conn_fd);
+			sigprocmask(SIG_UNBLOCK, &sigchildset, NULL);
+			continue;
+		}
+	}
+
 	/* Fork a sub-process to handle the connection. */
 	r = fork();
 	if ( r < 0 )
@@ -846,6 +935,26 @@
 	    handle_request();
 	    exit( 0 );
 	    }
+	    
+	currproc++;
+	
+	if (maxperip != 0) {
+		int i;
+		
+		/* record in table of clients */
+		for (i = 0; i < maxproc; i++) {
+			if (clients[i].cpid == 0) {
+				clients[i].cpid = r;
+				clients[i].caddr = usa.sa_in.sin_addr.s_addr;
+				break;
+			}
+		}
+		
+		if (i == maxproc)
+			syslog(LOG_CRIT, "client connection table full!");
+	}
+	sigprocmask(SIG_UNBLOCK, &sigchildset, NULL);
+	
 	(void) close( conn_fd );
 	}
     }
@@ -855,9 +964,9 @@
 usage( void )
     {
 #ifdef USE_SSL
-    (void) fprintf( stderr, "usage:  %s [-C configfile] [-D] [-S] [-E certfile] [-Y cipher] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage]\n", argv0 );
+    (void) fprintf( stderr, "usage:  %s [-C configfile] [-D] [-S] [-E certfile] [-Y cipher] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage] [-maxproc max_concurrent_procs] [-maxperip max_concurrent_procs_per_ip] [-cpelement path host]\n", argv0 );
 #else /* USE_SSL */
-    (void) fprintf( stderr, "usage:  %s [-C configfile] [-D] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage]\n", argv0 );
+    (void) fprintf( stderr, "usage:  %s [-C configfile] [-D] [-p port] [-d dir] [-dd data_dir] [-c cgipat] [-u user] [-h hostname] [-r] [-v] [-l logfile] [-i pidfile] [-T charset] [-P P3P] [-M maxage] [-maxproc max_concurrent_procs] [-maxperip max_concurrent_procs_per_ip] [-cpelement path host]\n", argv0 );
 #endif /* USE_SSL */
     exit( 1 );
     }
@@ -1128,7 +1237,7 @@
     char* cp;
     int r, file_len, i;
     const char* index_names[] = {
-	"index.html", "index.htm", "index.xhtml", "index.xht", "Default.htm",
+	"index.php", "index.html", "index.htm", "index.xhtml", "index.xht", "Default.htm",
 	"index.cgi" };
 
     /* Set up the timeout for reading. */
@@ -1166,9 +1275,11 @@
     ** solution is writev() (as used in thttpd), or send the headers with
     ** send(MSG_MORE) (only available in Linux so far).
     */
+	/*
     r = 1;
     (void) setsockopt(
 	conn_fd, IPPROTO_TCP, TCP_NOPUSH, (void*) &r, sizeof(r) );
+	*/
 #endif /* TCP_NOPUSH */
 
 #ifdef USE_SSL
@@ -1215,11 +1326,13 @@
 	send_error( 400, "Bad Request", "", "Can't parse request." );
     *protocol++ = '\0';
     protocol += strspn( protocol, " \t\012\015" );
-    query = strchr( path, '?' );
-    if ( query == (char*) 0 )
-	query = "";
-    else
-	*query++ = '\0';
+    if (!captivemode) {
+	    query = strchr( path, '?' );
+	    if ( query == (char*) 0 )
+			query = "";
+	    else
+			*query++ = '\0';
+	}
 
     /* Parse the rest of the request headers. */
     while ( ( line = get_request_line() ) != (char*) 0 )
@@ -1286,6 +1399,81 @@
 	method = METHOD_POST;
     else
 	send_error( 501, "Not Implemented", "", "That method is not implemented." );
+	
+	if (captivemode) {
+		/* only accept GET in captive portal mode */
+		int iscpelement = 0;
+		
+		captive_reqpath = path;
+		
+		if (cpelementpath != NULL && cpelementhost != NULL &&
+				host != NULL && strcmp(cpelementhost, host) == 0) {
+			/* the host name in the request headers matches our host name;
+			   see if the request matches a CP element */
+			char *mypath, *myfile;
+			
+			mypath = e_strdup(path);
+			
+			strdecode(mypath, mypath);
+			if (mypath[0] == '/') {
+				myfile = &(mypath[1]);
+				de_dotdot( myfile );
+				
+				iscpelement = 1;
+				
+				/* any slashes left? */
+				for (i = 0; myfile[i]; i++) {
+					if (myfile[i] == '/') {
+						iscpelement = 0;
+						break;
+					}
+				}
+				
+				if (iscpelement && myfile[0] != '\0' &&
+					!(myfile[0] == '.' && myfile[1] == '.' &&
+				   		myfile[2] == '\0')) {
+					
+					char *cpelpath;
+					
+					iscpelement = 0;
+					
+					/* see if that CP element exists */
+					cpelpath = e_malloc(strlen(myfile) + strlen(cpelementpath) + 2);
+					
+					strcpy(cpelpath, cpelementpath);
+					strcat(cpelpath, "/");
+					strcat(cpelpath, myfile);
+					
+					r = stat(cpelpath, &sb);
+					if (r == 0 && !S_ISDIR(sb.st_mode)) {
+						iscpelement = 1;
+						file = cpelpath;
+						path = mypath;
+						pathinfo = 0;
+					}
+				} else {
+					iscpelement = 0;
+				}
+			}
+		}
+		
+    	/* Set up the timeout for writing. */
+#ifdef HAVE_SIGSET
+		(void) sigset( SIGALRM, handle_write_timeout );
+#else /* HAVE_SIGSET */
+		(void) signal( SIGALRM, handle_write_timeout );
+#endif /* HAVE_SIGSET */
+		(void) alarm( WRITE_TIMEOUT );
+		
+		if (iscpelement) {
+			do_file();
+		} else {
+			path = "/index.php";
+			file = "index.php";
+			do_cgi();
+		}
+		
+	} else {
 
     strdecode( path, path );
     if ( path[0] != '/' )
@@ -1360,7 +1548,7 @@
 
 	got_one: ;
 	}
-
+	}
 #ifdef USE_SSL
     SSL_free( ssl );
 #endif /* USE_SSL */
@@ -2117,6 +2305,7 @@
     int envn;
     char* cp;
     char buf[256];
+    char rp[MAXPATHLEN];
 
     envn = 0;
     envp[envn++] = build_env( "PATH=%s", CGI_PATH );
@@ -2135,6 +2324,7 @@
     envp[envn++] = build_env(
 	"REQUEST_METHOD=%s", get_method_str( method ) );
     envp[envn++] = build_env( "SCRIPT_NAME=%s", path );
+    envp[envn++] = build_env( "SCRIPT_FILENAME=%s", realpath(file, rp) );
     if ( pathinfo != (char*) 0 )
 	{
 	envp[envn++] = build_env( "PATH_INFO=/%s", pathinfo );
@@ -2166,6 +2356,9 @@
 	envp[envn++] = build_env( "AUTH_TYPE=%s", "Basic" );
     if ( getenv( "TZ" ) != (char*) 0 )
 	envp[envn++] = build_env( "TZ=%s", getenv( "TZ" ) );
+	
+	if (captive_reqpath != NULL)
+		envp[envn++] = build_env("CAPTIVE_REQPATH=%s", captive_reqpath);
 
     envp[envn] = (char*) 0;
     return envp;
@@ -2341,8 +2534,6 @@
 
     send_error_body( s, title, text );
 
-    send_error_tail();
-
     send_response();
 
 #ifdef USE_SSL
@@ -2378,14 +2569,15 @@
     /* Send built-in error page. */
     buflen = snprintf(
 	buf, sizeof(buf), "\
-<HTML>\n\
-<HEAD><TITLE>%d %s</TITLE></HEAD>\n\
-<BODY BGCOLOR=\"#cc9999\" TEXT=\"#000000\" LINK=\"#2020ff\" VLINK=\"#4040cc\">\n\
-<H4>%d %s</H4>\n",
+<html>\n\
+<head><title>%d %s</title></head>\n\
+<body>\n\
+<h3>%d %s</h3>\n",
 	s, title, s, title );
     add_to_response( buf, buflen );
     buflen = snprintf( buf, sizeof(buf), "%s\n", text );
     add_to_response( buf, buflen );
+    send_error_tail();
     }
 
 
@@ -2416,7 +2608,7 @@
     {
     char buf[500];
     int buflen;
-
+/*
     if ( match( "**MSIE**", useragent ) )
 	{
 	int n;
@@ -2430,13 +2622,10 @@
 	buflen = snprintf( buf, sizeof(buf), "-->\n" );
 	add_to_response( buf, buflen );
 	}
-
+*/
     buflen = snprintf( buf, sizeof(buf), "\
-<HR>\n\
-<ADDRESS><A HREF=\"%s\">%s</A></ADDRESS>\n\
-</BODY>\n\
-</HTML>\n",
-	SERVER_URL, SERVER_SOFTWARE );
+</body>\n\
+</html>\n");
     add_to_response( buf, buflen );
     }
 
@@ -2457,8 +2646,10 @@
     start_response();
     buflen = snprintf( buf, sizeof(buf), "%s %d %s\015\012", protocol, status, title );
     add_to_response( buf, buflen );
+/*
     buflen = snprintf( buf, sizeof(buf), "Server: %s\015\012", SERVER_SOFTWARE );
     add_to_response( buf, buflen );
+*/
     now = time( (time_t*) 0 );
     (void) strftime( timebuf, sizeof(timebuf), rfc1123_fmt, gmtime( &now ) );
     buflen = snprintf( buf, sizeof(buf), "Date: %s\015\012", timebuf );
@@ -3034,8 +3225,10 @@
     {
     /* Don't need to set up the handler again, since it's a one-shot. */
 
+	/*
     syslog( LOG_NOTICE, "exiting due to signal %d", sig );
     (void) fprintf( stderr, "%s: exiting due to signal %d\n", argv0, sig );
+	*/
     closelog();
     exit( 1 );
     }
@@ -3096,6 +3289,23 @@
 		}
 	    break;
 	    }
+	currproc-- ;
+	
+	if (maxperip != 0) {
+		int i;
+		
+		/* remove from list of clients */
+		for (i = 0; i < maxproc; i++) {
+			if (clients[i].cpid == pid) {
+				clients[i].cpid = 0;
+				break;
+			}
+		}
+		
+		if (i == maxproc)
+			syslog(LOG_CRIT, "reaped child %d not found in table!", pid);
+	}
+	
 	}
 
     /* Restore previous errno. */
@@ -3128,7 +3338,9 @@
 static void
 handle_read_timeout( int sig )
     {
+	/*
     syslog( LOG_INFO, "%.80s connection timed out reading", ntoa( &client_addr ) );
+	*/
     send_error(
 	408, "Request Timeout", "",
 	"No request appeared within a reasonable time period." );
