# $Id: asterisk_makefile.patch 72 2007-04-26 15:59:03Z michael.iedema $
# patch for AskoziaPBX

--- asterisk/Makefile.orig	Wed Mar 14 10:41:41 2007
+++ asterisk/Makefile	Wed Mar 14 10:44:22 2007
@@ -83,7 +83,7 @@
 OVERWRITE=y
 
 # Include debug and macro symbols in the executables (-g) and profiling info (-pg)
-DEBUG=-g3
+#DEBUG=-g3
 
 # Staging directory
 # Files are copied here temporarily during the install process
