--- asterisk/apps/app_voicemail.c.orig	Mon Jan 28 21:42:43 2008
+++ asterisk/apps/app_voicemail.c	Fri Feb  8 08:45:52 2008
@@ -2088,7 +2088,7 @@
 	return res;
 }
 
-static int invent_message(struct ast_channel *chan, struct ast_vm_user *vmu, char *ext, int busy, char *ecodes)
+static int invent_message(struct ast_channel *chan, struct ast_vm_user *vmu, char *ext, int busy, char *ecodes, char *readnum)
 {
 	int res;
 	char fn[PATH_MAX];
@@ -2107,7 +2107,7 @@
 		res = ast_stream_and_wait(chan, "vm-theperson", chan->language, ecodes);
 		if (res)
 			return res;
-		res = ast_say_digit_str(chan, ext, ecodes, chan->language);
+		res = ast_say_digit_str(chan, readnum, ecodes, chan->language);
 	}
 
 	if (res)
@@ -2798,7 +2798,7 @@
 	signed char record_gain;
 };
 
-static int leave_voicemail(struct ast_channel *chan, char *ext, struct leave_vm_options *options)
+static int leave_voicemail(struct ast_channel *chan, char *ext, struct leave_vm_options *options, char *readnum)
 {
 #ifdef IMAP_STORAGE
 	int newmsgs, oldmsgs;
@@ -2912,7 +2912,7 @@
 			/* The file did not exist */
 			if (option_debug)
 				ast_log(LOG_DEBUG, "%s doesn't exist, doing what we can\n", prefile);
-			res = invent_message(chan, vmu, ext, ast_test_flag(options, OPT_BUSY_GREETING), ecodes);
+			res = invent_message(chan, vmu, ext, ast_test_flag(options, OPT_BUSY_GREETING), ecodes, readnum);
 		}
 		if (res < 0) {
 			if (option_debug)
@@ -4107,7 +4107,7 @@
 		/* Send VoiceMail */
 		memset(&leave_options, 0, sizeof(leave_options));
 		leave_options.record_gain = record_gain;
-		cmd = leave_voicemail(chan, mailbox, &leave_options);
+		cmd = leave_voicemail(chan, mailbox, &leave_options, mailbox);
 	} else {
 
 		/* Forward VoiceMail */
@@ -6958,6 +6958,7 @@
 	AST_DECLARE_APP_ARGS(args,
 		AST_APP_ARG(argv0);
 		AST_APP_ARG(argv1);
+		AST_APP_ARG(argv2); /* added new option to hold "readback number" */
 	);
 
 	u = ast_module_user_add(chan);
@@ -6970,7 +6971,7 @@
 	if (!ast_strlen_zero(data)) {
 		tmp = ast_strdupa(data);
 		AST_STANDARD_APP_ARGS(args, tmp);
-		if (args.argc == 2) {
+		if (args.argc >= 2) {
 			if (ast_app_parse_options(vm_app_options, &flags, opts, args.argv1)) {
 				ast_module_user_remove(u);
 				return -1;
@@ -7028,7 +7029,7 @@
 		args.argv0 = ast_strdupa(tmp);
 	}
 
-	res = leave_voicemail(chan, args.argv0, &leave_options);
+	res = leave_voicemail(chan, args.argv0, &leave_options, args.argv2);
 
 	if (res == ERROR_LOCK_PATH) {
 		ast_log(LOG_ERROR, "Could not leave voicemail. The path is already locked.\n");
@@ -8164,7 +8165,7 @@
 				
 				memset(&leave_options, 0, sizeof(leave_options));
 				leave_options.record_gain = record_gain;
-				res = leave_voicemail(chan, mailbox, &leave_options);
+				res = leave_voicemail(chan, mailbox, &leave_options, mailbox);
 				if (!res)
 					res = 't';
 				ast_config_destroy(msg_cfg);
