--- asterisk-1.6.1.11/Makefile.rules.orig	2009-12-01 16:34:42.000000000 +0100
+++ asterisk-1.6.1.11/Makefile.rules	2009-12-01 16:39:32.000000000 +0100
@@ -124,6 +124,6 @@
 
 %: %.o
 	$(ECHO_PREFIX) echo "   [LD] $^ -> $@"
-	$(CMD_PREFIX) $(CXX) $(STATIC_BUILD) -o $@ $(PTHREAD_CFLAGS) $(_ASTLDFLAGS) $^ $(CXX_LIBS) $(ASTLDFLAGS)
+	$(CMD_PREFIX) $(CC) $(STATIC_BUILD) -o $@ $(PTHREAD_CFLAGS) $(_ASTLDFLAGS) $^ $(CC_LIBS) $(ASTLDFLAGS)
 
 dist-clean:: clean
