--- dahdi-linux-2.3.0.1/drivers/dahdi/wcb4xxp/wcb4xxp.h.orig	2010-08-12 11:17:41.000000000 +0200
+++ dahdi-linux-2.3.0.1/drivers/dahdi/wcb4xxp/wcb4xxp.h	2010-08-12 11:28:26.000000000 +0200
@@ -426,7 +426,11 @@
 	BN4S0,		/* Beronet BN4S0			*/
 	BN8S0,		/* BeroNet BN8S0			*/
 	BSWYX_SX2,	/* Swyx 4xS0 SX2 QuadBri		*/
-	QUADBRI_EVAL	/* HFC-4S CCD Eval. Board		*/
+	QUADBRI_EVAL,	/* HFC-4S CCD Eval. Board		*/
+	GENERIC1,	/* cards for which we have ids but no confirmed tests on led function, etc */
+	GENERIC2,
+	GENERIC4,
+	GENERIC8
 	};
 
 /* This structure exists one per card */
--- dahdi-linux-2.3.0.1/drivers/dahdi/wcb4xxp/base.c.orig	2010-08-12 11:05:39.000000000 +0200
+++ dahdi-linux-2.3.0.1/drivers/dahdi/wcb4xxp/base.c	2010-08-12 11:25:17.000000000 +0200
@@ -139,6 +139,21 @@
 static struct devtype hfc4s_SW = {"Swyx 4xS0 SX2 QuadBri", .ports = 4, .card_type = BSWYX_SX2 };
 static struct devtype hfc4s_EV = {"CCD HFC-4S Eval. Board", .ports = 4,
 					.card_type = QUADBRI_EVAL };
+static struct devtype BERONET_BN1S0_MINI =  {"BeroNet BN1S0 miniPCI", .ports = 1, .card_type = GENERIC1 };
+static struct devtype BERONET_BN2S0_MINI =  {"BeroNet BN2S0 miniPCI", .ports = 2, .card_type = GENERIC2 };
+static struct devtype BERONET_BN2S0_PCIE =  {"BeroNet BN2S0 PCIe", .ports = 2, .card_type = GENERIC2 };
+static struct devtype BERONET_BN4S0_MINI =  {"BeroNet BN4S0 miniPCI", .ports = 4, .card_type = GENERIC4 };
+static struct devtype BERONET_BN4S0_PCIE =  {"BeroNet BN4S0 PCIe", .ports = 4, .card_type = GENERIC4 };
+static struct devtype BERONET_BN8S0_PLUS =  {"BeroNet BN8S0+", .ports = 8, .card_type = GENERIC8 };
+static struct devtype JUNGHANNS_DUO_MINI =  {"Junghanns.NET duoBRI miniPCI", .ports = 2, .card_type = GENERIC2 };
+static struct devtype JUNGHANNS_QUAD_MINI = {"Junghanns.NET quadBRI miniPCI", .ports = 4, .card_type = GENERIC4 };
+static struct devtype JUNGHANNS_QUAD_PCI2 = {"Junghanns.NET quadBRI 2.0 PCI", .ports = 4, .card_type = GENERIC4 };
+static struct devtype JUNGHANNS_OCTO_PCI =  {"Junghanns.NET octoBRI PCI", .ports = 8, .card_type = GENERIC8 };
+static struct devtype GENERIC_HFC4S_1 =     {"HFC-4S", .ports = 4, .card_type = GENERIC4 };
+static struct devtype GENERIC_HFC8S_1 =     {"HFC-8S IOB4ST Recording", .ports = 8, .card_type = GENERIC8 };
+static struct devtype GENERIC_HFC8S_2 =     {"HFC-8S IOB8ST", .ports = 8, .card_type = GENERIC8 };
+static struct devtype GENERIC_HFC8S_3 =     {"HFC-8S", .ports = 8, .card_type = GENERIC8 };
+static struct devtype GENERIC_HFC8S_4 =     {"HFC-8S", .ports = 8, .card_type = GENERIC8 };
  
 #define CARD_HAS_EC(card) ((card)->card_type == B410P)
 
@@ -428,6 +443,7 @@
 	mb();
 
 	switch (b4->card_type) {
+	case GENERIC8: /* fall through */
 	case OCTOBRI:  /* fall through */
 	case B800P_OV: /* fall through */
 	case BN8S0:
@@ -1047,6 +1063,7 @@
 	unsigned long irq_flags;
 
 	switch (b4->card_type) {
+	case GENERIC8: /* fall through */
 	case B800P_OV: /* fall through */
 	case OCTOBRI: /* fall through */
 	case BN8S0:
@@ -2900,6 +2917,16 @@
 	{ 0x1397, 0x16b8, 0x1397, 0xb56b, 0, 0, (unsigned long)&hfc8s_BN },
 	{ 0x1397, 0x08b4, 0x1397, 0xb540, 0, 0, (unsigned long)&hfc4s_SW },
 	{ 0x1397, 0x08b4, 0x1397, 0x08b4, 0, 0, (unsigned long)&hfc4s_EV },
+	{ 0x1397, 0x08b4, 0x1397, 0xb567, 0, 0, (unsigned long)&BERONET_BN1S0_MINI },   /* BeroNet BN1S0 miniPCI */
+	{ 0x1397, 0x08b4, 0x1397, 0xb569, 0, 0, (unsigned long)&BERONET_BN2S0_MINI },   /* BeroNet BN2S0 miniPCI */
+	{ 0x1397, 0x08b4, 0x1397, 0xb568, 0, 0, (unsigned long)&BERONET_BN4S0_MINI },   /* BeroNet BN4S0 miniPCI */
+	{ 0x1397, 0x08b4, 0x1397, 0xb762, 0, 0, (unsigned long)&BERONET_BN4S0_PCIE },   /* BeroNet BN4S0 PCIe */
+	{ 0x1397, 0x08b4, 0x1397, 0xb559, 0, 0, (unsigned long)&JUNGHANNS_DUO_MINI },   /* Junghanns.NET duoBRI miniPCI */
+	{ 0x1397, 0x08b4, 0x1397, 0xb558, 0, 0, (unsigned long)&JUNGHANNS_QUAD_MINI },  /* Junghanns.NET quadBRI miniPCI */
+	{ 0x1397, 0x08b4, 0x1397, 0xb620, 0, 0, (unsigned long)&GENERIC_HFC4S_1 },      /* HFC-4S */
+	{ 0x1397, 0x16b8, 0x1397, 0xb521, 0, 0, (unsigned long)&GENERIC_HFC8S_1 },      /* HFC-8S IOB4ST Recording */
+	{ 0x1397, 0x16b8, 0x1397, 0xb522, 0, 0, (unsigned long)&GENERIC_HFC8S_2 },      /* HFC-8S IOB8ST */
+	{ 0x1397, 0x16b8, 0x1397, 0xb622, 0, 0, (unsigned long)&GENERIC_HFC8S_4 },      /* HFC-8S */
 	{0, }
 
 };
--- dahdi-linux-2.3.0.1/drivers/dahdi/dahdi_transcode.c	2011-06-30 13:41:34.000000000 +0200
+++ dahdi-linux-2.3.0.1/drivers/dahdi/dahdi_transcode.c.2	2011-07-05 01:27:55.000000000 +0200
@@ -46,7 +46,7 @@
  * is used as a simplistic way to spread the load amongst the different hardware
  * transcoders in the system. */
 static LIST_HEAD(active_list);
-static spinlock_t translock = SPIN_LOCK_UNLOCKED;
+static DEFINE_SPINLOCK(translock);
 
 EXPORT_SYMBOL(dahdi_transcoder_register);
 EXPORT_SYMBOL(dahdi_transcoder_unregister);
