--- asterisk/channels/chan_sip.c.orig	2008-06-16 12:22:14.000000000 +0200
+++ asterisk/channels/chan_sip.c	2008-06-16 13:15:12.000000000 +0200
@@ -1359,6 +1359,8 @@
 static int sip_show_users(int fd, int argc, char *argv[]);
 static int _sip_show_peers(int fd, int *total, struct mansession *s, const struct message *m, int argc, const char *argv[]);
 static int sip_show_peers(int fd, int argc, char *argv[]);
+static int _sip_show_statuses(int fd, int *total, struct mansession *s, const struct message *m, int argc, const char *argv[]);
+static int sip_show_statuses(int fd, int argc, char *argv[]);
 static int sip_show_objects(int fd, int argc, char *argv[]);
 static void  print_group(int fd, ast_group_t group, int crlf);
 static const char *dtmfmode2str(int mode) attribute_const;
@@ -10094,6 +10096,44 @@
 #undef FORMAT2
 }
 
+/*! \brief  CLI Show Statuses command */
+static int sip_show_statuses(int fd, int argc, char *argv[])
+{
+	return _sip_show_statuses(fd, NULL, NULL, NULL, argc, (const char **) argv);
+}
+
+/*! \brief  _sip_show_statuses: Execute sip show statuses command */
+static int _sip_show_statuses(int fd, int *total, struct mansession *s, const struct message *m, int argc, const char *argv[])
+{
+#define FORMAT  "%s %s\n"
+
+	char name[256];
+
+	switch (argc) {
+	case 3:
+		break;
+	default:
+		return RESULT_SHOWUSAGE;
+	}
+	
+	ASTOBJ_CONTAINER_TRAVERSE(&peerl, 1, do {
+		char status[20] = "";
+		char srch[2000];
+		char pstatus;
+		
+		ASTOBJ_RDLOCK(iterator);
+		ast_copy_string(name, iterator->name, sizeof(name));
+		pstatus = peer_status(iterator, status, sizeof(status));
+		snprintf(srch, sizeof(srch), FORMAT, name, status);
+		ast_cli(fd, FORMAT, name, status);
+		ASTOBJ_UNLOCK(iterator);
+
+	} while(0) );
+
+	return RESULT_SUCCESS;
+#undef FORMAT
+}
+
 /*! \brief List all allocated SIP Objects (realtime or static) */
 static int sip_show_objects(int fd, int argc, char *argv[])
 {
@@ -11721,6 +11761,10 @@
 "       Shows all details on one SIP peer and the current status.\n"
 "       Option \"load\" forces lookup of peer in realtime storage.\n";
 
+static char show_statuses_usage[] = 
+"Usage: sip show statuses\n"
+"       Lists all peers statuses.\n";
+
 static char prune_realtime_usage[] =
 "Usage: sip prune realtime [peer|user] [<name>|all|like <pattern>]\n"
 "       Prunes object(s) from the cache.\n"
@@ -18054,6 +18098,10 @@
 	sip_show_peers, "List defined SIP peers",
 	show_peers_usage },
 
+	{ { "sip", "show", "statuses", NULL },
+	sip_show_statuses, "List SIP peers statuses",
+	show_statuses_usage },
+
 	{ { "sip", "show", "registry", NULL },
 	sip_show_registry, "List SIP registration status",
 	show_reg_usage },
