--- asterisk-1.6.1.11/main/asterisk.c.orig	2009-12-07 18:45:28.000000000 +0100
+++ asterisk-1.6.1.11/main/asterisk.c	2009-12-07 18:46:22.000000000 +0100
@@ -1052,7 +1052,7 @@
 				break;
 		}
 	}
-	ast_verb(3, "Remote UNIX connection disconnected\n");
+	/*ast_verb(3, "Remote UNIX connection disconnected\n");*/
 	close(con->fd);
 	close(con->p[0]);
 	close(con->p[1]);
@@ -1115,8 +1115,9 @@
 				fdprint(s, "No more connections allowed\n");
 				ast_log(LOG_WARNING, "No more connections allowed\n");
 				close(s);
-			} else if (consoles[x].fd > -1) 
-				ast_verb(3, "Remote UNIX connection\n");
+			} else if (consoles[x].fd > -1) {
+				/*ast_verb(3, "Remote UNIX connection\n");*/
+			}
 		}
 	}
 	return NULL;
