Index: chan_gsm.c
===================================================================
--- chan_gsm.c	(revision 1558)
+++ chan_gsm.c	(working copy)
@@ -1,6 +1,14 @@
+#include <asterisk.h>
+
+#if defined(__NetBSD__) || defined(__FreeBSD__)
+#include <pthread.h>
+#include <signal.h>
+#else
+#include <sys/signal.h>
+#endif
+
 #include <stdio.h>
 #include <stdlib.h>
-#include <pthread.h>
 #include <semaphore.h>
 #include <string.h>
 #include <sys/socket.h>
@@ -12,12 +20,11 @@
 #include <fcntl.h>
 #include <sys/ioctl.h>
 #include <sys/file.h>
-#include <signal.h>
 #include <sys/types.h>
 #include <sys/ipc.h>
 #include <sys/msg.h>
 
-#include <asterisk.h>
+#include <asterisk/lock.h>
 #include <asterisk/module.h>
 #include <asterisk/channel.h>
 #include <asterisk/config.h>
@@ -43,7 +50,7 @@
 #include <asterisk/csel.h>
 
 #include "gsmctl.h"
-#include "gsm_version.h"
+//#include "gsm_version.h"
 void* load_module_thread(void* data);
 
 ast_mutex_t gsm_mutex;
@@ -1356,7 +1363,8 @@
 		return NULL;
 	}
 
-	ast_cli(a->fd,"\nchan_gsm version (" CHAN_GSM_VERSION ")\n");
+	//ast_cli(a->fd,"\nchan_gsm version (" CHAN_GSM_VERSION ")\n");
+	ast_cli(a->fd,"\nchan_gsm version (trunk)\n");
 	return CLI_SUCCESS;
 }
 
Index: compat/res_statemachine.c
===================================================================
--- compat/res_statemachine.c	(revision 1558)
+++ compat/res_statemachine.c	(working copy)
@@ -24,9 +24,9 @@
  */
 
 
+#include "asterisk/compat-16.h"
 
 
-
 #include "asterisk/statemachine.h"
 #include "asterisk/module.h"
 #include "asterisk/logger.h"
Index: compat/res_csel.c
===================================================================
--- compat/res_csel.c	(revision 1558)
+++ compat/res_csel.c	(working copy)
@@ -24,9 +24,9 @@
  */
 
 
+#include "asterisk/compat-16.h"
 
 
-
 #include "asterisk/csel.h"
 #include "asterisk/module.h"
 #include "asterisk/logger.h"
Index: compat/res_configman.c
===================================================================
--- compat/res_configman.c	(revision 1558)
+++ compat/res_configman.c	(working copy)
@@ -32,9 +32,9 @@
  */
 
 
+#include "asterisk/compat-16.h"
 
 
-
 #include <errno.h>
 #include <stdarg.h>
 #include <stdio.h>
