--- fonulator-2.0.1/Makefile.in.orig	2010-04-19 14:37:21.000000000 +0200
+++ fonulator-2.0.1/Makefile.in	2010-04-19 14:39:34.000000000 +0200
@@ -54,13 +54,7 @@
 	tokens.$(OBJEXT) status.$(OBJEXT) dsp.$(OBJEXT) \
 	error.$(OBJEXT) flash.$(OBJEXT) dlist.$(OBJEXT)
 fonulator_OBJECTS = $(am_fonulator_OBJECTS)
-ifeq (x86_64,$(ARCH))
-fonulator_DEPENDENCIES = @LIBOBJS@ /usr/lib64/libnet.a \
-        /usr/lib64/libpcap.a /usr/lib/libargtable2.a
-else
-fonulator_DEPENDENCIES = @LIBOBJS@ /usr/lib/libnet.a \
-	/usr/lib/libpcap.a /usr/lib/libargtable2.a
-endif
+fonulator_DEPENDENCIES = @LIBOBJS@
 DEFAULT_INCLUDES = -I.@am__isrc@
 depcomp = $(SHELL) $(top_srcdir)/depcomp
 am__depfiles_maybe = depfiles
@@ -184,11 +178,7 @@
 man_MANS = fonulator.1
 fonulator_SOURCES = fonulator.c keys.c tokens.l status.c dsp.c error.c flash.c dlist.c
 noinst_HEADERS = config.h dsp.h error.h fonulator.h state.h status.h tokens.h tree.h ver.h dlist.h
-ifeq (x86_64,$(ARCH))
-fonulator_LDADD = @LIBOBJS@ @LIBFB@ /usr/lib64/libnet.a /usr/lib64/libpcap.a /usr/lib/libargtable2.a
-else
-fonulator_LDADD = @LIBOBJS@ @LIBFB@ /usr/lib/libnet.a /usr/lib/libpcap.a /usr/lib/libargtable2.a 
-endif
+fonulator_LDADD = @LIBOBJS@ @LIBFB@ -lnet -lpcap -largtable2 
 EXTRA_DIST = $(man_MANS)
 all: config.h
 	$(MAKE) $(AM_MAKEFLAGS) all-am
