--- ntpclient-2007/ntpclient.c.orig	2009-10-06 13:46:59.000000000 +0200
+++ ntpclient-2007/ntpclient.c	2009-10-06 13:47:26.000000000 +0200
@@ -146,7 +146,7 @@
 #ifdef __linux__
 	struct timex txc;
 	txc.modes=0;
-	if (__adjtimex(&txc) < 0) {
+	if (adjtimex(&txc) < 0) {
 		perror("adjtimex"); exit(1);
 	}
 	return txc.freq;
@@ -163,7 +163,7 @@
 	struct timex txc;
 	txc.modes = ADJ_FREQUENCY;
 	txc.freq = new_freq;
-	if (__adjtimex(&txc) < 0) {
+	if (adjtimex(&txc) < 0) {
 		perror("adjtimex"); exit(1);
 	}
 	return txc.freq;
