--- busybox/networking/httpd.c.old	2009-09-05 19:14:32.000000000 +0200
+++ busybox/networking/httpd.c	2009-09-08 21:06:50.000000000 +0200
@@ -2101,8 +2105,15 @@ static void handle_incoming_and_exit(con
 		}
 		send_cgi_and_exit(urlcopy, prequest, length, cookie, content_type);
 	}
+	//if (prequest != request_GET && prequest != request_HEAD) {
+	//	send_headers_and_exit(HTTP_NOT_IMPLEMENTED);
+	//}
+#endif
+
+	if (urlp[-1] == '/')
+		strcpy(urlp, index_page);
+	if (stat(tptr, &sb) == 0) {
 #if ENABLE_FEATURE_HTTPD_CONFIG_WITH_SCRIPT_INTERPR
-	{
 		char *suffix = strrchr(tptr, '.');
 		if (suffix) {
 			Htaccess *cur;
@@ -2112,16 +2123,7 @@ static void handle_incoming_and_exit(con
 				}
 			}
 		}
-	}
 #endif
-	if (prequest != request_GET && prequest != request_HEAD) {
-		send_headers_and_exit(HTTP_NOT_IMPLEMENTED);
-	}
-#endif  /* FEATURE_HTTPD_CGI */
-
-	if (urlp[-1] == '/')
-		strcpy(urlp, index_page);
-	if (stat(tptr, &sb) == 0) {
 		file_size = sb.st_size;
 		last_mod = sb.st_mtime;
 	}
